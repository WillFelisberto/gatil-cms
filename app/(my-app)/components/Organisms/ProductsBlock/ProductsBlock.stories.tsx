import type { Meta, StoryObj } from '@storybook/react';

import type { BlockProps } from '@/app/(my-app)/types/block-props';

import { ProductsBlock } from './ProductsBlock';

const meta: Meta<typeof ProductsBlock> = {
  title: 'Organisms/ProductsBlock',
  component: ProductsBlock,
  tags: ['autodocs']
};

export default meta;
type Story = StoryObj<typeof ProductsBlock>;

export const Default: Story = {
  args: {
    blockType: 'products',
    title: 'Ajude o Gatil comprando ❤️',
    products: [
      {
        name: 'Caneca Personalizada',
        description: 'Ilustração exclusiva. Ajude um resgatado!',
        price: 49.9,
        image: {
          url: '/gato.jpg',
          alt: 'Caneca do gatil'
        },
        buyLink: 'https://www.mercadopago.com.br/caneca'
      },
      {
        name: 'Chaveiro Resgatado',
        description: 'Fofo demais para não ter um.',
        price: 24.9,
        image: {
          url: '/gato.jpg',
          alt: 'Chaveiro solidário'
        },
        buyLink: 'https://www.mercadopago.com.br/chaveiro'
      }
    ]
  } as BlockProps<'products'>
};
