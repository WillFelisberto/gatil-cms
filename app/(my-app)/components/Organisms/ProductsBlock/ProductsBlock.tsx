'use client';

import type { BlockProps } from '@/app/(my-app)/types/block-props';
import { Media } from '@/payload-types';

import { ProductCard } from '../../Molecules/ProductCard';

export const ProductsBlock = (props: BlockProps<'products'>) => {
  const { title, products } = props;

  return (
    <section data-testid="products-block" className="w-full px-4 py-16 bg-white">
      <div className="max-w-7xl mx-auto">
        <h2
          className="text-3xl md:text-4xl font-bold text-center text-[#013274] mb-12"
          data-testid="products-title"
        >
          {title}
        </h2>

        <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 gap-10">
          {products &&
            products?.map((product, i) => (
              <ProductCard
                key={product.buyLink || i}
                name={product.name}
                description={product.description || ''}
                price={product.price}
                image={product.image as Media}
                buyLink={product.buyLink}
              />
            ))}
        </div>
      </div>
    </section>
  );
};
