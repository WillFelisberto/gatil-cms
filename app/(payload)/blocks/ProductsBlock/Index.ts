import { Block } from 'payload';
import { NumberField } from '@nouance/payload-better-fields-plugin/Number';

export const ProductsBlock: Block = {
  slug: 'products',
  labels: {
    singular: 'Produto solidário',
    plural: 'Produtos solidários'
  },
  fields: [
    {
      name: 'title',
      type: 'text',
      label: 'Título da seção',
      defaultValue: 'Compre e ajude!',
      required: true
    },
    {
      name: 'products',
      type: 'array',
      label: 'Lista de Produtos',
      fields: [
        {
          name: 'name',
          type: 'text',
          label: 'Nome do produto',
          required: true
        },
        {
          name: 'description',
          type: 'textarea',
          label: 'Descrição'
        },
        ...NumberField(
          {
            name: 'price',
            label: 'Preço (R$)',
            required: true
          },
          {
            prefix: 'R$ ',
            thousandSeparator: '.',
            decimalSeparator: ',',
            decimalScale: 2,
            fixedDecimalScale: true
          }
        ),
        {
          name: 'image',
          type: 'upload',
          relationTo: 'media',
          required: true
        },
        {
          name: 'buyLink',
          type: 'text',
          label: 'Link do Mercado Pago',
          required: true
        }
      ]
    }
  ]
};
