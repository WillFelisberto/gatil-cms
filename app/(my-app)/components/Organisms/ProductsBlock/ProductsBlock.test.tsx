import type { BlockProps } from '@/app/(my-app)/types/block-props';
import { render, screen } from '@/tests/test-utils';

import { ProductsBlock } from './ProductsBlock';

const mockProductsBlock: BlockProps<'products'> = {
  blockType: 'products',
  title: 'Produtos Solidários',
  products: [
    {
      name: 'Caneca do Gatil',
      description: 'Ajude um resgatado com essa caneca linda.',
      price: 39.9,
      buyLink: 'https://www.mercadopago.com.br/caneca',
      image: {
        createdAt: '2023-01-01',
        updatedAt: '2023-01-01',
        id: 1,
        url: '/caneca.jpg',
        alt: 'Caneca solidária'
      }
    },
    {
      name: 'Chaveiro Fofo',
      description: 'Leve um pedacinho do Gatil com você.',
      price: 19.9,
      buyLink: 'https://www.mercadopago.com.br/chaveiro',
      image: {
        createdAt: '2023-01-01',
        updatedAt: '2023-01-01',
        id: 2,
        url: '/chaveiro.jpg',
        alt: 'Chaveiro solidário'
      }
    }
  ]
};

describe('<ProductsBlock />', () => {
  it('renders section and title', () => {
    const { container } = render(<ProductsBlock {...mockProductsBlock} />);
    expect(screen.getByTestId('products-block')).toBeInTheDocument();
    expect(screen.getByTestId('products-title')).toHaveTextContent('Produtos Solidários');

    expect(container).toMatchSnapshot();
  });

  it('renders all product cards', () => {
    render(<ProductsBlock {...mockProductsBlock} />);
    const cards = screen.getAllByTestId('product-card');
    expect(cards).toHaveLength(2);
  });

  it('renders correct product names', () => {
    render(<ProductsBlock {...mockProductsBlock} />);
    expect(screen.getByText('Caneca do Gatil')).toBeInTheDocument();
    expect(screen.getByText('Chaveiro Fofo')).toBeInTheDocument();
  });
});
