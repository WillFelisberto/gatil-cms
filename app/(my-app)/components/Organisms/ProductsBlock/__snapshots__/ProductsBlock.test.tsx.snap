// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsBlock /> renders section and title 1`] = `
<div>
  <section
    class="w-full px-4 py-16 bg-gray-50"
    data-testid="products-block"
  >
    <div
      class="max-w-7xl mx-auto"
    >
      <h2
        class="text-3xl md:text-4xl font-bold text-center text-[#013274] mb-12"
        data-testid="products-title"
      >
        Produtos Solidários
      </h2>
      <div
        class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 gap-10"
      >
        <div
          class="bg-white rounded-2xl shadow-lg overflow-hidden flex flex-col"
          data-testid="product-card"
        >
          <div
            class="relative w-full h-64"
          >
            <img
              alt="Caneca solidária"
              class="object-cover"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              sizes="(max-width: 768px) 100vw, (max-width: 1024px) 50vw, 33vw"
              src="/_next/image?url=%2Fcaneca.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fcaneca.jpg&w=256&q=75 256w, /_next/image?url=%2Fcaneca.jpg&w=384&q=75 384w, /_next/image?url=%2Fcaneca.jpg&w=640&q=75 640w, /_next/image?url=%2Fcaneca.jpg&w=750&q=75 750w, /_next/image?url=%2Fcaneca.jpg&w=828&q=75 828w, /_next/image?url=%2Fcaneca.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fcaneca.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fcaneca.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fcaneca.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fcaneca.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="p-6 flex-1 flex flex-col justify-between gap-4"
          >
            <div>
              <h3
                class="text-xl font-semibold text-[#013274] mb-1"
                data-testid="product-name"
              >
                Caneca do Gatil
              </h3>
              <p
                class="text-sm text-gray-700"
                data-testid="product-description"
              >
                Ajude um resgatado com essa caneca linda.
              </p>
            </div>
            <div
              class="flex flex-col gap-2"
            >
              <span
                class="text-lg font-bold text-pink-600"
                data-testid="product-price"
              >
                R$ 39,90
              </span>
              <a
                class="inline-block text-center bg-pink-600 hover:bg-pink-700 text-white font-medium py-2 px-4 rounded-xl transition-all"
                data-testid="product-cta"
                href="https://www.mercadopago.com.br/caneca"
                rel="noopener noreferrer"
                target="_blank"
              >
                Comprar
              </a>
            </div>
          </div>
        </div>
        <div
          class="bg-white rounded-2xl shadow-lg overflow-hidden flex flex-col"
          data-testid="product-card"
        >
          <div
            class="relative w-full h-64"
          >
            <img
              alt="Chaveiro solidário"
              class="object-cover"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              sizes="(max-width: 768px) 100vw, (max-width: 1024px) 50vw, 33vw"
              src="/_next/image?url=%2Fchaveiro.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fchaveiro.jpg&w=256&q=75 256w, /_next/image?url=%2Fchaveiro.jpg&w=384&q=75 384w, /_next/image?url=%2Fchaveiro.jpg&w=640&q=75 640w, /_next/image?url=%2Fchaveiro.jpg&w=750&q=75 750w, /_next/image?url=%2Fchaveiro.jpg&w=828&q=75 828w, /_next/image?url=%2Fchaveiro.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fchaveiro.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fchaveiro.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fchaveiro.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fchaveiro.jpg&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
          <div
            class="p-6 flex-1 flex flex-col justify-between gap-4"
          >
            <div>
              <h3
                class="text-xl font-semibold text-[#013274] mb-1"
                data-testid="product-name"
              >
                Chaveiro Fofo
              </h3>
              <p
                class="text-sm text-gray-700"
                data-testid="product-description"
              >
                Leve um pedacinho do Gatil com você.
              </p>
            </div>
            <div
              class="flex flex-col gap-2"
            >
              <span
                class="text-lg font-bold text-pink-600"
                data-testid="product-price"
              >
                R$ 19,90
              </span>
              <a
                class="inline-block text-center bg-pink-600 hover:bg-pink-700 text-white font-medium py-2 px-4 rounded-xl transition-all"
                data-testid="product-cta"
                href="https://www.mercadopago.com.br/chaveiro"
                rel="noopener noreferrer"
                target="_blank"
              >
                Comprar
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
